### PR TITLE
Added picture in picture possible feature to iOS flat video renderer

### PR DESCRIPTION
**Pull request by pavliuko**

### CODE REVIEW

File: VideoRenderer/VideoRenderer/VideoStreamViewController.swift [hunk 1]
```diff
@@ -208,12 +211,21 @@ public final class VideoStreamViewController: UIViewController, RendererProtocol
...
                         return pipController
                     } else {
-                        guard
-                            let layer = videoView?.playerLayer,
-                            props.pictureInPictureActive else { return nil }
+                        guard let layer = videoView?.playerLayer else { return nil }
                         let pipController = AVPictureInPictureController(playerLayer: layer)
                         pipController?.delegate = self
                         self.pictureInPictureController = pipController
+                        if let pictureInPictureController = self.pictureInPictureController {
```
Comment (AndriiMoskvin) [inline VideoRenderer/VideoRenderer/VideoStreamViewController.swift]: You have `pipController` above - what is the reason for this `if let`?

File: VideoRenderer/VideoRenderer/PictureInPictureControllerObserver.swift [hunk 2]
```diff
@@ -0,0 +1,59 @@
+//  Copyright © 2017 One by Aol : Publishers. All rights reserved.
+
+import Foundation
+import AVKit
+
+public final class PictureInPictureControllerObserver: NSObject {
```
Comment (AndriiMoskvin) [inline VideoRenderer/VideoRenderer/PictureInPictureControllerObserver.swift:6]: Use `SystemPlayerObserver` as an example of implementing KVO tracking.

File: VideoRenderer/VideoRenderer/PictureInPictureControllerObserver.swift [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    
+    public struct Callbacks {
+        public var possibilityChanged: Action<(Bool)>?
+    }
+    
+    private let callbacks: Callbacks
+    private var removeObserver: (PictureInPictureControllerObserver) -> ()
+    
+    private struct Context {
```
Comment (AndriiMoskvin) [inline VideoRenderer/VideoRenderer/PictureInPictureControllerObserver.swift]: Replace with `#keyPath`.

File: VideoRenderer/VideoRenderer/PictureInPictureControllerObserver.swift [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        
+        guard let change = change else { fatalError("Change should not be nil!") }
+        guard let context = context else { fatalError("Added observer without context!") }
+        
+        guard context == &Context.pictureInPicturePossible else {
+            super.observeValue(forKeyPath: keyPath, of: object, change: change, context: context)
+            return
+        }
+        
+        let newValue = change[NSKeyValueChangeKey.newKey] as! NSNumber
```
Comment (AndriiMoskvin) [inline VideoRenderer/VideoRenderer/PictureInPictureControllerObserver.swift]: Do not use force-casts in production code.

File: VideoRenderer/VideoRenderer/PictureInPictureControllerObserver.swift [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        let isPictureInPicturePossible: Bool = newValue.boolValue
```
Comment (AndriiMoskvin) [inline VideoRenderer/VideoRenderer/PictureInPictureControllerObserver.swift]: `Bool` could be dropped - type inference system will calculate the type.

File: VideoRenderer/VideoRenderer/PictureInPictureControllerObserver.swift (hunk 2 above)
Comment (AndriiMoskvin) [reply VideoRenderer/VideoRenderer/PictureInPictureControllerObserver.swift:6]: `PlayerObserver` is a bit outdated - I would like to migrate to `SystemPlayerObserver` in nearest future.

**Review by AndriiMoskvin — CHANGES_REQUESTED** [review]

File: VideoRenderer/VideoRenderer/VideoStreamViewController.swift (hunk 1 above)
Comment (pavliuko) [reply VideoRenderer/VideoRenderer/VideoStreamViewController.swift]: `pipController` is optional, that's why we have verification here. If you worrying about bad code readability, I propose to replace with: `if let pipController = pipController {`

File: VideoRenderer/VideoRenderer/PictureInPictureControllerObserver.swift [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        
+        guard keyPath == keyPathPiPPossible else {
+            super.observeValue(forKeyPath: keyPath, of: object, change: change, context: context)
+            return
+        }
+        
+        guard let newValue = change[NSKeyValueChangeKey.newKey] as? NSNumber else { return }
+        let isPictureInPicturePossible = newValue.boolValue
+        
+        emit(.didChangedPossibility(to: isPictureInPicturePossible))
```
Comment (AndriiMoskvin) [inline VideoRenderer/VideoRenderer/PictureInPictureControllerObserver.swift]: `emit(.didChangedPossibility(to: newValue.boolValue))` - is ok

File: VideoRenderer/VideoRenderer/PictureInPictureControllerObserver.swift [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        
+        guard let keyPath = keyPath else { fatalError("Unexpected nil keypath!") }
+        guard let change = change else { fatalError("Change should not be nil!") }
+        
+        guard keyPath == keyPathPiPPossible else {
+            super.observeValue(forKeyPath: keyPath, of: object, change: change, context: context)
+            return
+        }
+        
+        guard let newValue = change[NSKeyValueChangeKey.newKey] as? NSNumber else { return }
```
Comment (AndriiMoskvin) [inline VideoRenderer/VideoRenderer/PictureInPictureControllerObserver.swift]: `.newKey`

Comment (AndriiMoskvin) [reply VideoRenderer/VideoRenderer/PictureInPictureControllerObserver.swift]: `NSKeyValueChangeKey` could be dropped.

File: VideoRenderer/VideoRenderer/VideoStreamViewController.swift [hunk 8]
```diff
@@ -208,12 +211,19 @@ public final class VideoStreamViewController: UIViewController, RendererProtocol
                     if let pipController = self.pictureInPictureController as? AVPictureInPictureController {
                         return pipController
                     } else {
-                        guard
-                            let layer = videoView?.playerLayer,
-                            props.pictureInPictureActive else { return nil }
+                        guard let layer = videoView?.playerLayer else { return nil }
                         let pipController = AVPictureInPictureController(playerLayer: layer)
```
Comment (AndriiMoskvin) [inline VideoRenderer/VideoRenderer/VideoStreamViewController.swift]: Maybe just have `guard let pipController = AVPictureInPictureController(playerLayer: layer)`.
Then you don't need `if let` at the bottom.

**Review by AndriiMoskvin — CHANGES_REQUESTED** [review]

File: VideoRenderer/VideoRenderer/VideoStreamViewController.swift [hunk 9]
```diff
@@ -210,10 +213,14 @@ public final class VideoStreamViewController: UIViewController, RendererProtocol
                     } else {
                         guard
                             let layer = videoView?.playerLayer,
-                            props.pictureInPictureActive else { return nil }
-                        let pipController = AVPictureInPictureController(playerLayer: layer)
-                        pipController?.delegate = self
+                            let pipController = AVPictureInPictureController(playerLayer: layer) else { return nil }
+                        pipController.delegate = self
                         self.pictureInPictureController = pipController
```
Comment (AndriiMoskvin) [inline VideoRenderer/VideoRenderer/VideoStreamViewController.swift]: `self` could be dropped.

**Review by AndriiMoskvin — APPROVED** [review]

**Pull request merged**